pennylane/optimize/rotosolve.py

```diff
@@ -14,78 +14,122 @@
-        _assert_integer(num_freqs)
-        num_freqs = [num_freqs] * sum(requires_grad)
-        num_freqs_flat = num_freqs
-    else:
-        num_freqs_flat = []
-        for num_frequency in num_freqs:
-            if np.isscalar(num_frequency):
-                _assert_integer(num_frequency)
-                num_frequency_flat = num_frequency
-            else:
-                num_frequency_flat = list(_flatten(num_frequency))
-                for _num_freq in num_frequency_flat:
-                    _assert_integer(_num_freq)
-                num_frequency_flat = np.array(num_frequency_flat, dtype=int)
-            num_freqs_flat.append(num_frequency_flat)
-        if len(num_freqs) != sum(requires_grad):
-            raise ValueError(
-                "The length of the provided numbers of frequencies "
-                f"({len(num_freqs)}) does not match the number of function arguments "
-                f"({sum(requires_grad)})."
-            )
-    return num_freqs, num_freqs_flat
+import pennylane as qml
 
 
 def _brute_optimizer(fun, num_steps, **kwargs):
     r"""Brute force optimizer, wrapper of scipy.optimizer.brute that repeats it
-    ``num_steps`` times. Signature is as expected by ``RotosolveOptimizer._rotosolve``
-    below, providing a scalar minimal position and the function value at that position."""
+    ``num_steps`` times. Signature is as expected by ``RotosolveOptimizer._min_numeric``
+    below, returning a scalar minimal position and the function value at that position."""
     width = 2 * np.pi
     x_min = 0.0
     Ns = kwargs.pop("Ns")
     for _ in range(num_steps):
         _range = (x_min - width / 2, x_min + width / 2)
         x_min, y_min, *_ = brute(fun, ranges=(_range,), full_output=True, Ns=Ns, **kwargs)
+        # We only ever use this function for 1D optimization
+        x_min = x_min[0]
         width /= Ns
 
     return x_min, y_min
 
 
 def _shgo_optimizer(fun, **kwargs):
     r"""Wrapper for ``scipy.optimize.shgo`` (Simplicial Homology global optimizer).
-    Signature is as expected by ``RotosolveOptimizer._rotosolve`` below, providing
+    Signature is as expected by ``RotosolveOptimizer._min_numeric`` below, providing
     a scalar minimal position and the function value at that position."""
     opt_res = shgo(fun, **kwargs)
-    return opt_res.x, opt_res.fun
+    return opt_res.x[0], opt_res.fun
+
+
+def _prepare_optimizer(optimizer, optimizer_kwargs):
+    r"""Set default optimizer and optimizer keyword arguments
+    for the one-dimensional optimization in each substep of Rotosolve."""
+    optimizer_kwargs = optimizer_kwargs or {}
+    optimizer = optimizer or "brute"
+
+    if optimizer == "brute":
+        optimizer = _brute_optimizer
+        optimizer_kwargs.setdefault("num_steps", 4)
+        optimizer_kwargs.setdefault("Ns", 100)
+    elif optimizer == "shgo":
+        optimizer = _shgo_optimizer
+        optimizer_kwargs.setdefault("bounds", ((-np.pi, np.pi),))
```

Comment:
would this cause issues for gates that are not periodic in `[-π, π]`?

Reply:
Yes, it might. In that case users will have to pass the bounds themselves. Note that the "very default" behaviour is to choose "brute", so that an entirely blind user will not experience this problem. Once they start to change `optimizer`, the `bounds` should be considered.
I'll add a warning to the doc string regarding this.

Reply:
I took a look again, the domain `[-π, π]` actually was baked into the `brute` wrapper as well - and in an immutable way for the user! I changed this and for both, `shgo` and `brute`, `ranges` is now a valid keyword argument and I added the warning.
Also, the default range now considers the smallest frequency and is set to `[-π/f_min, π/f_min]`.
In general I noticed that determining the required domain to optimize over is not straight-forward, and setting the interval per parameter is not readily possible for users...

Reply:
nice :)